File: src/PebView.php
<?php

// 严格模式
declare(strict_types=1);

use \FFI\CData;

/**
 * PebView类
 */
class PebView
{
    private CData $pv;
    private \FFI $ffi;

    /**
     * 构造函数
     *
     * @param boolean $debug 是否开启调试模式，默认false
     * @param CData|null $window 窗口句柄，默认null
     * @throws \Exception 加载库失败时抛出异常
     */
    public function __construct(bool $debug = false, CData|null $window = null)
    {
        $header = file_get_contents(__DIR__ . DIRECTORY_SEPARATOR . "PebView.h");
        try {
            $this->ffi = \FFI::cdef($header, $this->getLibFile());
        } catch (\Exception $e) {
            throw new \Exception("Failed to load PebView library: {$e->getMessage()}");
        }
        $this->pv = $this->ffi->webview_create($debug ? 1 : 0, $window);
    }

    /**
     * 销毁PebView实例
     *
     * @return self
     */
    public function destroy(): self
    {
        $this->getError($this->ffi->webview_destroy($this->pv));
        return $this;
    }

    /**
     * 运行PebView实例
     *
     * @return self
     */
    public function run(): self
    {
        $this->getError($this->ffi->webview_run($this->pv));
        return $this;
    }

    /**
     * 终止PebView实例
     *
     * @return self
     */
    public function terminate(): self
    {
        $this->getError($this->ffi->webview_terminate($this->pv));
        return $this;
    }

    /**
     * 调度任务
     * 一般用于在其他线程中调用PebView实例的方法
     *
     * @return self
     */
    public function dispatch(callable $callable): self
    {
        $c_fn = function ($v, $d) use ($callable) {
            $callable($this);
        };
        $this->getError($this->ffi->webview_dispatch($this->pv, $c_fn, null));
        return $this;
    }

    /**
     * 获取窗口句柄
     *
     * @return CData 窗口句柄
     */
    public function getWindow(): CData
    {
        return $this->ffi->webview_get_window($this->pv);
    }

    /**
     * 获取原生句柄
     *
     * @param integer $kind 原生句柄类型，0表示窗口句柄，1表示小部件，2表示浏览器
     * @return CData 原生句柄
     */
    public function getNativeHandle(int $kind): CData
    {
        return $this->ffi->webview_get_native_handle($this->pv, $kind);
    }

    /**
     * 设置标题
     *
     * @param string $title 标题
     * @return self
     */
    public function setTitle(string $title): self
    {
        $this->getError($this->ffi->webview_set_title($this->pv, $title));
        return $this;
    }

    /**
     * 设置大小
     *
     * @param integer $width 宽度
     * @param integer $height 高度
     * @param integer $hints 0表示不指定，1表示最小化，2表示最大化，3表示固定大小
     * @return self
     */
    public function setSize(int $width, int $height, int $hints = 0): self
    {
        $this->getError($this->ffi->webview_set_size($this->pv, $width, $height, $hints));
        return $this;
    }

    /**
     * 导航到指定URL
     *
     * @param string $url URL
     * @return self
     */
    public function navigate(string $url): self
    {
        $this->getError($this->ffi->webview_navigate($this->pv, $url));
        return $this;
    }

    /**
     * 加载html到内容
     *
     * @param string $html html字符串
     * @return self
     */
    public function setHtml(string $html): self
    {
        $this->getError($this->ffi->webview_set_html($this->pv, $html));
        return $this;
    }

    /**
     * 注入JavaScript代码在加载页面时立即执行
     * 代码将在 window.onload之前执行
     *
     * @param string $js
     * @return self
     */
    public function init(string $js): self
    {
        $this->getError($this->ffi->webview_init($this->pv, $js));
        return $this;
    }

    /**
     * 立即执行JavaScript代码
     *
     * @param string $js JavaScript代码
     * @return self
     */
    public function eval(string $js): self
    {
        $this->getError($this->ffi->webview_eval($this->pv, $js));
        return $this;
    }

    /**
     * 绑定JavaScript函数
     *
     * @param string $name 函数名
     * @param Closure $function 回调函数
     * @return self
     */
    public function bind(string $name, Closure $function): self
    {
        $c_fn = function ($id, $req, $arg) use ($function) {
            $res = $function($this, $id, json_decode($req, true), $arg);
            if ($res && (is_object($res) || is_array($res))) {
                $this->getError($this->ffi->webview_return($this->pv, $id, 0, json_encode($res, 320)));
            }
        };
        $this->getError($this->ffi->webview_bind($this->pv, $name, $c_fn, null));
        return $this;
    }

    /**
     * 卸载JavaScript函数
     *
     * @param string $name 函数名
     * @return self
     */
    public function unbind(string $name): self
    {
        $this->getError($this->ffi->webview_unbind($this->pv, $name));
        return $this;
    }

    /**
     * 响应JavaScript函数调用
     *
     * @param CData $id 函数调用ID
     * @param integer $status 0表示成功，其他值表示失败
     * @param array|object $result 响应结果，必须是数组或对象
     * @return self
     */
    public function value(CData $id, int $status, array|object $result): self
    {
        $this->getError($this->ffi->webview_return($this->pv, $id, $status, json_encode($result, 320)));
        return $this;
    }

    /**
     * 获取版本信息
     *
     * @return CData
     */
    public function version(): CData
    {
        return $this->ffi->webview_version();
    }

    public function alert(string $message): self
    {
        $this->ffi->osdialog_message(0, 0, $message);
        return $this;
    }

    /**
     * 获取错误信息
     *
     * @param integer $code 错误码
     */
    private function getError(int $code): void
    {
        switch ($code) {
            case -5:
                throw new \Exception("窗口未找到");
            case -4:
                throw new \Exception("操作已被用户取消");
            case -3:
                throw new \Exception("无效状态检测到");
            case -2:
                throw new \Exception("一个或多个无效参数已指定，例如在函数调用中。");
            case -1:
                throw new \Exception("发生了未指定的错误。可能需要更多特定的错误代码");
            case 1:
                throw new \Exception("内存分配失败");
            case 2:
                throw new \Exception("操作失败");
            default:
        }
    }

    /**
     * 获取库文件路径
     *
     * @return string
     */
    private function getLibFile(): string
    {
        $os = PHP_OS_FAMILY;
        switch ($os) {
            case "Windows":
                $suffix = ".dll";
                break;
            case "Darwin":
                $suffix = ".dylib";
                break;
            case "Linux":
                $suffix = ".so";
                break;
            default:
                throw new \Exception("Unsupported OS: {$os}");
        }
        $libFile = dirname(__DIR__)  . DIRECTORY_SEPARATOR
            . "lib" . DIRECTORY_SEPARATOR
            . $os . DIRECTORY_SEPARATOR
            . "PebView{$suffix}";
        return $libFile;
    }
}

File: source/pebview.cpp
#include "php++.h"
#include "webview.h"
#include "icon.h"
#include "window.h"
#include "osdialog.h"
#include "toast.h"
#include <map>

// Window API ------------------------------------------------------------------

// 定义 窗口提示 枚举
PhpEnum(WindowHint, t_int)
    EnumCase(WindowHint, None, 0)
        EnumCase(WindowHint, Min, 1)
            EnumCase(WindowHint, Max, 2)
                EnumCase(WindowHint, Fixed, 3);

webview_t wb;
void *tray;
// 关闭回调（webview_set_close_callback 无 userdata 槽位，用全局变量桥接）
static zval *close_cb = nullptr;
static zval close_this;

// 托盘菜单回调桥接（tray_menu::callback 无 userdata 槽位，用 id 映射）
struct TrayCb
{
    zval *cb;
    zval self;
};
static std::map<int, TrayCb> tray_cb_map;

// C 回调入口：ptr 指向 tray_menu（首个字段是 int id），查找映射 → 调用 PHP
static void TrayCbBrig(const void *ptr)
{
    int id = *(int *)ptr;
    auto it = tray_cb_map.find(id);
    if (it != tray_cb_map.end() && it->second.cb)
    {
        zval retval = p_func(it->second.cb).add(&it->second.self).call();
        zval_ptr_dtor(&retval);
    }
}

// 定义Window类
Class(Window);

/**
 * @brief 定义构造函数
 *
 * @param bool debug 是否开启调试模式
 */
Method(Window, __construct, t_void,
       Arg(t_bool, "debug"))
{
    // 构造函数
    bool debug = GetArgBool(1);
    wb = webview_create(debug, NULL);
}

/**
 * @brief 定义销毁窗口函数
 *
 */
Method(Window, __destruct, t_void)
{
    // 销毁窗口
    webview_destroy(wb);
}

/**
 * @brief 定义运行窗口函数
 *
 * @return t_Object
 */
Method(Window, run, t_object)
{
    // 运行窗口
    webview_run(wb);
    RetThis();
}

/**
 * @brief 定义终止窗口函数
 *
 */
Method(Window, terminate, t_void)
{
    window_tray_remove(tray);
    // 终止窗口
    webview_terminate(wb);
}

/**
 * @brief 定义设置窗口图标函数
 * @param string iconPath 图标文件路径
 * @return t_Object
 */
Method(Window, setIcon, t_object,
       Arg(t_str, "iconPath"))
{
    const char *iconPath = GetArgStr(1);
    enum SetIconErrorCode code = set_icon(webview_get_window(wb), iconPath);
    if (code == WINDOW_NOT_FOUND)
    {
        PhpError("Window not found");
    }
    else if (code == ICON_NOT_FOUND)
    {
        PhpError("Icon file not found");
    }
    else if (code == OS_UNSUPPORTED)
    {
        PhpError("Operating system not supported");
    }
    RetThis();
}

/**
 * @brief 定义设置窗口标题函数
 *
 * @param string title 窗口标题
 * @return t_Object
 */
Method(Window, setTitle, t_object,
       Arg(t_str, "title"))
{
    const char *title = GetArgStr(1);
    webview_set_title(wb, title);
    RetThis();
}

/**
 * @brief 定义设置窗口大小函数
 *
 * @param int width 窗口宽度
 * @param int height 窗口高度
 * @return t_Object
 */
Method(Window, setSize, t_object,
       Arg(t_int, "width")
           Arg(t_int, "height")
               ArgEnum(WindowHint, "hint"))
{
    int width = GetArgInt(1);
    int height = GetArgInt(2);
    int hint = GetArgEnumIntDfl(3, 0);
    webview_hint_t hint_val;
    if (hint == 0)
    {
        hint_val = WEBVIEW_HINT_NONE;
    }
    else if (hint == 1)
    {
        hint_val = WEBVIEW_HINT_MIN;
    }
    else if (hint == 2)
    {
        hint_val = WEBVIEW_HINT_MAX;
    }
    else if (hint == 3)
    {
        hint_val = WEBVIEW_HINT_FIXED;
    }
    else
    {
        hint_val = WEBVIEW_HINT_NONE;
    }
    webview_set_size(wb, width, height, hint_val);
    RetThis();
}

/**
 * @brief 初始化窗口
 *
 * @param string js 初始化js代码
 */
Method(Window, init, t_object,
       Arg(t_str, "js"))
{
    webview_init(wb, GetArgStr(1));
    RetThis();
}

/**
 * @brief 定义执行js代码函数
 *
 * @param string js js代码
 * @return t_Object
 */
Method(Window, eval, t_object,
       Arg(t_str, "js"))
{
    webview_eval(wb, GetArgStr(1));
    RetThis();
}

/**
 * @brief 定义设置html内容函数
 *
 * @param string html html内容
 * @return t_Object
 */
Method(Window, setHtml, t_object,
       Arg(t_str, "html"))
{
    const char *html = GetArgStr(1);
    webview_set_html(wb, html);
    RetThis();
}

/**
 * @brief 定义导航函数
 *
 * @param string url 导航url
 * @return t_Object
 */
Method(Window, navigate, t_object,
       Arg(t_str, "url"))
{
    const char *url = GetArgStr(1);
    webview_navigate(wb, url);
    RetThis();
}

/**
 * @brief 定义绑定js函数
 *
 * @param string name 绑定函数名称
 * @param callable callback PHP回调函数，接收一个参数（JS传来的JSON字符串），返回值会被传回JS
 * @return t_Object
 */
Method(Window, bind, t_object,
       Arg(t_str, "name")
           Arg(t_func, "callback"))
{
    const char *name = GetArgStr(1);
    zval *cb = PhpCbDup(_C(2));

    webview_bind(wb, name, [](const char *s, const char *r, void *a)
                 {
                    // ① 解码 JS 传来的 JSON 参数
                    zval decoded;
                    PhpCall("json_decode", &decoded, r ? r : "", true);

                    // ② 调用 PHP 回调: callback($seq, $decoded)，拿到返回值
                    zval retval = p_func((zval *)a).add(&decoded).call();
                    zval_ptr_dtor(&decoded);

                    // ③ 返回值统一 json_encode，传给 JS
                    zval json;
                    PhpCall("json_encode", &json, &retval);
                    zval_ptr_dtor(&retval);

                    webview_return(wb, s, 0,
                        Z_TYPE(json) == t_str ? Z_STRVAL(json) : "");
                    zval_ptr_dtor(&json); }, cb);

    RetThis();
}

/**
 * @brief 解绑js函数
 *
 * @param string name 绑定函数名称
 * @return t_Object
 */
Method(Window, unBind, t_object,
       Arg(t_str, "name"))
{
    const char *name = GetArgStr(1);
    webview_unbind(wb, name);
    RetThis();
}

/**
 * @brief 定义设置窗口关闭回调函数
 *
 * @param callable callback PHP回调函数，返回值为true时允许窗口关闭，false时阻止窗口关闭
 * @return t_Object
 */
Method(Window, setCloseCallback, t_object,
       Arg(t_func, "callback"))
{
    // 释放旧回调
    if (close_cb)
    {
        zval_ptr_dtor(close_cb);
    }
    close_cb = PhpCbDup(_C(1));

    // 存储 $this（close 回调触发时 getThis() 不可用）
    zval_ptr_dtor(&close_this);
    ZVAL_COPY(&close_this, getThis());

    webview_set_close_callback(wb, [](void * /*unused, always nullptr*/) -> int
                               {
        if (!close_cb) return 0;
        zval retval = p_func(close_cb).add(&close_this).call();
        int result = (Z_TYPE(retval) == IS_TRUE) ? 1 : 0;  // true=关闭, false=阻止
        zval_ptr_dtor(&retval);
        return result; });
    RetThis();
}

/**
 * @brief 定义显示窗口函数
 * @return t_Object
 */
Method(Window, show, t_object)
{
    window_show(webview_get_window(wb));
    RetThis();
}

/**
 * @brief 定义隐藏窗口函数
 * @return t_Object
 */
Method(Window, hide, t_object)
{
    window_hide(webview_get_window(wb));
    RetThis();
}

/**
 * @brief 定义创建托盘函数
 *
 * @param string icon 托盘图标路径
 * @return t_Object
 */
Method(Window, tray, t_object,
       Arg(t_str, "icon"))
{
    const char *icon = GetArgStr(1);
    tray = window_tray(wb, icon);
    RetThis();
}

/**
 * @brief 定义添加托盘菜单函数
 *
 * @arg array menus 菜单项数组，每项为关联数组，支持键：text（字符串）、id（整数，可选）、disabled（布尔值，可选）、checked（布尔值，可选）、cb（PHP回调函数，可选）
 */
Method(Window, trayMenu, t_object,
       Arg(t_array, "menus"))
{
    p_arr arr = GetArgArr(1);
    zval *self = getThis();
    static int idx = 1;

    arr.values([&](zval *val)
               {
        if (PType(val) != IS_ARRAY) return;
        HashTable* ht = Z_ARRVAL_P(val);

        struct tray_menu menu = {0};
        menu.id = idx++;

        zval* zv;
        if ((zv = zend_hash_str_find(ht, "text",  4)) && Z_TYPE_P(zv) == IS_STRING) menu.text     = (char*)Z_STRVAL_P(zv);
        if ((zv = zend_hash_str_find(ht, "id",    2)) && Z_TYPE_P(zv) == IS_LONG)   menu.id       = (int)Z_LVAL_P(zv);
        if ((zv = zend_hash_str_find(ht, "disabled", 8)) && Z_TYPE_P(zv) == IS_TRUE)   menu.disabled = 1;
        if ((zv = zend_hash_str_find(ht, "checked",  7)) && Z_TYPE_P(zv) == IS_TRUE)   menu.checked  = 1;

        if ((zv = zend_hash_str_find(ht, "cb", 2)) && Z_TYPE_P(zv) == IS_OBJECT) {
            auto& slot = tray_cb_map[menu.id];
            if (slot.cb) zval_ptr_dtor(slot.cb);
            zval_ptr_dtor(&slot.self);
            slot.cb   = PhpCbDup(zv);
            ZVAL_COPY(&slot.self, self);
            menu.callback = TrayCbBrig;
        }

        window_tray_add_menu(tray, &menu); });

    RetThis();
}

// Dialog API ------------------------------------------------------------------

PhpEnum(DialogLevel, t_int)
    EnumCase(DialogLevel, Info, 0)
        EnumCase(DialogLevel, Warning, 1)
            EnumCase(DialogLevel, Error, 2);

PhpEnum(DialogBtn, t_int)
    EnumCase(DialogBtn, Ok, 0)
        EnumCase(DialogBtn, OkCancel, 1)
            EnumCase(DialogBtn, YesNo, 2);

PhpEnum(FileAction, t_int)
    EnumCase(FileAction, Open, 0)
        EnumCase(FileAction, OpenDir, 1)
            EnumCase(FileAction, Save, 2);

Class(Dialog);
StaticMethod(Dialog, msg, t_bool,
             Arg(t_str, "message")
                 ArgEnum(DialogLevel, "level")
                     ArgEnum(DialogBtn, "buttons"))
{
    const char *message = GetArgStr(1);
    int level = GetArgEnumIntDfl(2, 0);
    int buttons = GetArgEnumIntDfl(3, 0);

    osdialog_message_level msg_level;
    switch (level)
    {
    case 0:
        msg_level = OSDIALOG_INFO;
        break;
    case 1:
        msg_level = OSDIALOG_WARNING;
        break;
    case 2:
        msg_level = OSDIALOG_ERROR;
        break;
    default:
        msg_level = OSDIALOG_INFO;
    }

    osdialog_message_buttons msg_buttons;
    switch (buttons)
    {
    case 0:
        msg_buttons = OSDIALOG_OK;
        break;
    case 1:
        msg_buttons = OSDIALOG_OK_CANCEL;
        break;
    case 2:
        msg_buttons = OSDIALOG_YES_NO;
        break;
    default:
        msg_buttons = OSDIALOG_OK;
    }
    bool result = osdialog_message(msg_level, msg_buttons, message) == 1;
    RetBool(result);
}

StaticMethod(Dialog, prompt, t_str,
             Arg(t_str, "message")
                 ArgEnum(DialogLevel, "level")
                     Arg(t_str, "text"))
{
    const char *message = GetArgStr(1);
    int level = GetArgEnumIntDfl(2, 0);
    const char *text = GetArgStrDfl(3, "");

    osdialog_message_level msg_level;
    switch (level)
    {
    case 0:
        msg_level = OSDIALOG_INFO;
        break;
    case 1:
        msg_level = OSDIALOG_WARNING;
        break;
    case 2:
        msg_level = OSDIALOG_ERROR;
        break;
    default:
        msg_level = OSDIALOG_INFO;
    }

    char *result = osdialog_prompt(msg_level, message, text);
    if (result)
    {
        RetStr(result);
        free(result);
    }
    else
    {
        RetStr("");
    }
}

StaticMethod(Dialog, file, t_str,
             Arg(t_str, "dir")
                 Arg(t_str, "filename")
                     ArgEnum(FileAction, "action")
                         Arg(t_str, "filters"))
{
    const char *dir = GetArgStr(1);
    const char *filename = GetArgStr(2);
    int action = GetArgEnumInt(3);
    const char *filters_str = GetArgStrDfl(4, NULL);

    osdialog_file_action file_action;
    switch (action)
    {
    case 0:
        file_action = OSDIALOG_OPEN;
        break;
    case 1:
        file_action = OSDIALOG_OPEN_DIR;
        break;
    case 2:
        file_action = OSDIALOG_SAVE;
        break;
    default:
        file_action = OSDIALOG_OPEN;
    }

    osdialog_filters *filters = nullptr;
    if (filters_str)
    {
        filters = osdialog_filters_parse(filters_str);
    }

    char *result = osdialog_file(file_action, dir, filename, filters);
    if (filters)
    {
        osdialog_filters_free(filters);
    }
    if (result)
    {
        RetStr(result);
        free(result);
    }
    else
    {
        RetStr("");
    }
}

// Toast API ------------------------------------------------------------------
Class(Toast);
StaticMethod(Toast, show, t_bool,
             Arg(t_str, "app")
                 Arg(t_str, "title")
                     Arg(t_str, "message")
                         Arg(t_str, "icon"))
{
    const char *app = GetArgStr(1);
    const char *title = GetArgStr(2);
    const char *message = GetArgStr(3);
    const char *icon = GetArgStrDfl(4, NULL);

    bool result = toastShow(app, title, message, icon);
    RetBool(result);
}

Module("pebview", "v0.0.1", "Kingbes\\PebView");